File: go/merge_sort/merge_sort_array.go
package merge_sort

// MergeSortArray should sort array with *stable* merge sort using provided comparator:
// comparator function should return negative value if arr[i]<arr[j], zero if arr[i]==arr[j]
// and positive value if arr[i]>arr[j]
func MergeSortArray[T any](arr []T, cmp func(T, T) int) {
	tmp := make([]T, len(arr))
	mergeSortArray(arr, tmp, cmp)
}

func mergeSortArray[T any](arr, tmp []T, cmp func(T, T) int) {
	n := len(arr)
	m := n / 2
	if m == 0 {
		return
	}

	mergeSortArray(arr[:m], tmp, cmp)
	mergeSortArray(arr[m:], tmp, cmp)

	i, j, k := 0, m, 0
	for ; i < m || j < n; k++ {
		if j == n || i < m && cmp(arr[i], arr[j]) <= 0 {
			tmp[k] = arr[i]
			i++
		} else {
			tmp[k] = arr[j]
			j++
		}
	}

	copy(arr, tmp)
}

File: python/merge_sort/merge_sort_array.py

# merge_sort_array should sort array with *stable* merge sort using provided comparator:
# comparator function should return negative value if arr[i]<arr[j], zero if arr[i]==arr[j]
# and positive value if arr[i]>arr[j]
def merge_sort_array(arr, cmp):
	# raise NotImplementedError('TODO')

	merge_sort_array_recurse(arr, arr.copy(), cmp, 0, len(arr))


def merge_sort_array_recurse(arr, tmp, cmp, l, r):
	if r-l <= 1:
		return
	m = (l+r)//2
	merge_sort_array_recurse(arr, tmp, cmp, l, m)
	merge_sort_array_recurse(arr, tmp, cmp, m, r)

	merge(arr, tmp, cmp, l, m, r)


def merge(arr, tmp, cmp, l, m, r):
	i, j, k = l, m, 0
	while i < m or j < r:
		if i < m and (j == r or cmp(arr[i], arr[j]) <= 0):
			tmp[k] = arr[i]
			i += 1
		else:
			tmp[k] = arr[j]
			j += 1
		k += 1
	arr[l:r] = tmp[:k]
